lib.php

```diff
@@ -0,0 +1,60 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Main library of plugin.
+ *
+ * @package    quizaccess_seb
+ * @author     Andrew Madden <andrewmadden@catalyst-au.net>
+ * @copyright  2019 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Serve a seb config file for a particular quiz.
+ *
+ * @param string $cmid The course module ID for a quiz with config.
+ * @param int $cachelifetime Time in seconds til cached file expires.
+ * @return bool Whether the file was created and served.
+ */
+function serve_quiz_config_xml(string $cmid, $cachelifetime = 300) {
+    // Check that the course module exists, user is logged into course and can access course module.
+    try {
+        // Try and get the course module.
+        $cm = get_coursemodule_from_id('quiz', $cmid, 0, false, MUST_EXIST);
+
+        // Make sure the user is logged in and has access to the module.
+        require_login($cm->course, false, $cm);
+    } catch (moodle_exception $e) {
+        debugging($e->getMessage(), DEBUG_DEVELOPER);
```

Comment:
probably don't catch it and just let moodle to sort it out as per usual

Reply:
If it's not caught, it will be echoed into the downloaded file and corrupt the config.

Reply:
then we need to throw it before like here https://github.com/catalyst/moodle-quizaccess_seb/pull/28/files#diff-9e5e2527d69c009a81b8ecd730f3957eR30

Reply:
is there any reason to wrap it up to the function?  are we going to reuse it? Maybe move everything to config.php and properly display error if something when wrong?

Reply:
For testing. Is it possible to unit test a script?

Because we change the headers that get built before the page loads any errors on the page will go into the download file. 

I can try and do a redirect if any errors occur maybe?

Reply:
ok. I see you covered it with the tests, which is good! But we don't have to output custom headers if something went wrong. It feels like this needs to be refactored to throw an exception if something goes wrong (not return false) and only if all good then do header things.

Reply:
That's much better. Refactored so headers only set up if config string is found and validated.